### PR TITLE
Check if sneaking for optimal speed in garden

### PR DESCRIPTION
**Pull request by lunaynx**
This is useful for people who use Fermento Boots with toggle sneak for wheat, carrot, potato and nether wart to achieve the desired speed, because the only way to get below 100 speed without Rancher's Boots is sneaking or a slowness potion.

I round down the speed here because generally when farming you rather want to overshoot your speed a bit than undershoot to ensure 20 BPS.

### CODE REVIEW

**Review by hannibal002 — CHANGES_REQUESTED** [review]
While I like this addition, I think we also need to show the current speed next to the wrong speed (if speed is wrong) to make the sneak impact more user friendly.

**Comment by lunaynx** [comment]
Yeah I will do that. I think we should also check to see if the player has been sneaking for some time already, otherwise it can trigger unnecessary alerts when they just sneak for a moment while standing still.

**Comment by lunaynx** [comment]
Done both.

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt [hunk 1]
```diff
@@ -15,18 +16,36 @@ import net.minecraft.client.gui.inventory.GuiEditSign
...
+        }
     private var optimalSpeed = -1
     private val currentSpeedPattern = " Speed: §r§f✦(?<speed>.*)".toPattern()
     private var lastWarnTime = 0L
     private var cropInHand: CropType? = null
     private var rancherOverlayList: List<List<Any?>> = emptyList()
 
+    @SubscribeEvent
+    fun onTick(event: LorenzTickEvent) {
+        if (sneaking) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt:42]: Maybe save the time stamp of the last non sneaking moment instead. (Mutable vs immutable state)

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt [hunk 2]
```diff
@@ -84,9 +103,14 @@ class GardenOptimalSpeed {
...
 
-        val text = "Optimal Speed: §f$optimalSpeed"
+        var text = "Optimal Speed: §f$optimalSpeed"
         if (optimalSpeed != currentSpeed) {
+            text += " (§eCurrent: §f$currentSpeed"
+            if (sneaking) text += " §7[Sneaking]"
+            text += "§f)"
+
             if (config.showOnHUD) config.pos.renderString("§c$text", posLabel = "Garden Optimal Speed")
+            if (sneaking && !sneakingPersistent) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt:113]: This logic means "no warning when sneaking for more than 5 seconds" and this cant be right. some loigcal error maybe?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
not sure

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt (hunk 2 above)
Comment (lunaynx) [reply src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt:113]: It means no warning if sneaking but not for long enough (5 seconds).

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/garden/GardenOptimalSpeed.kt:113]: ohh nvm

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**